sinks/hawkular/driver.go

```diff
@@ -0,0 +1,236 @@
+package hawkular
+
+import (
+	"fmt"
+	"net/url"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/GoogleCloudPlatform/heapster/extpoints"
+	"github.com/golang/glog"
+	"github.com/hawkular/hawkular-client-go/metrics"
+
+	sink_api "github.com/GoogleCloudPlatform/heapster/sinks/api"
+	kube_api "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+)
+
+const (
+	units_tag string = "units"
+	type_tag  string = "type"
+	start_tag string = "start_time"
+	desc_tag  string = "_description"
+	group_tag string = "group_id"
+	sep_str   string = "/"
+)
+
+type hawkularSink struct {
+	client  *metrics.Client
+	models  map[string]metrics.MetricDefinition // Model definitions
+	regLock sync.Mutex
+	reg     map[string]*metrics.MetricDefinition // Real definitions
+}
+
+// START: ExternalSink interface implementations
+
+func (self *hawkularSink) Register(mds []sink_api.MetricDescriptor) error {
+	self.regLock.Lock()
+	defer self.regLock.Unlock()
+
+	// Create model definitions based on the MetricDescriptors
+	for _, md := range mds {
+		hmd := self.descriptorToDefinition(&md)
+		self.models[md.Name] = hmd
+	}
+
+	// Fetch currently known metrics from Hawkular-Metrics and cache them
+	prev, err := self.client.Definitions(metrics.Gauge)
+	if err != nil {
+		return err
+	}
+
+	for _, p := range prev {
+		self.reg[p.Id] = p
```

Comment:
In this model, we would store the metric with a name: containerName/containerId/metricName and only the last one is available on the Register() phase. If the register would be called when new container is detected, then it would make sense to register here. But as we now want to create a metric per each container+metric combination it can't be done here, so that's why we store the model only here.